### PR TITLE
Sync stable-cadence branch with master

### PR DESCRIPTION
**Pull request by SupunS**
## Description

Sync stable-cadence branch with master
______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Comment by SupunS** [comment]
There were quite a few conflicts, mostly between the https://github.com/onflow/cadence/pull/2401, and the purity changes.

In addition to conflicts, had to do the following code changes:
- Fixing matchers (a4acb79d21eac0ca5912cf0fa6c90fd91cd09b15): 
  - There was a change to matchers (https://github.com/onflow/cadence/pull/2130) in the stable-cadence branch, but the newly added matchers was assuming the old behaviour.
- Fixing json encoding (cdeeabd0f13f9328642c04efba7027c7cce7b60b)
  - To accompany the changes went in https://github.com/onflow/cadence/pull/2401
- Updating reference invalidation (adca90bf340c63395eaa90f2f4a515bec6f005d7)
  - Updated the changes went in #2447 and #2453 to match the new purpose of invalidating the references.

File: encoding/ccf/decode.go [hunk 1]
```diff
@@ -2003,8 +2003,12 @@ func (d *Decoder) decodeFunctionTypeValue(visited *cadenceTypeByCCFTypeID) (cade
 		return nil, errors.New("unexpected nil function return type")
 	}
 
+	// TODO:
+	purity := cadence.FunctionPurityUnspecified
+
```
Comment (SupunS) [inline encoding/ccf/decode.go:2008]: https://github.com/onflow/cadence/issues/2457

File: encoding/ccf/encode.go [hunk 2]
```diff
@@ -61,7 +61,6 @@ func Encode(value cadence.Value) ([]byte, error) {
 	var w bytes.Buffer
 
 	enc := NewEncoder(&w)
-	defer enc.enc.Close()
```
Comment (turbolent) [inline encoding/ccf/encode.go]: Was this removal intentional?

File: runtime/sema/gen/main.go [hunk 3]
```diff
@@ -831,6 +839,16 @@ func functionTypeExpr(
 
 	var compositeElements []dst.Expr
 
+	if purityExpr != nil {
```
Comment (turbolent) [inline runtime/sema/gen/main.go:842]: Thanks for adding this!

File: docs/language/accounts.mdx [hunk 4]
```diff
@@ -148,19 +148,17 @@ pub struct AuthAccount {
...
     ///
     /// The public key must be encoded together with their signature algorithm, hashing algorithm and weight.
-    fun addPublicKey(_ publicKey: [UInt8])
+    pub fun addPublicKey(_ publicKey: [UInt8])
 
     /// **DEPRECATED**: Use `keys.revoke` instead.
     ///
     /// Revokes the key at the given index.
-    fun removePublicKey(_ index: Int)
+    pub fun removePublicKey(_ index: Int)
```
Comment (turbolent) [inline docs/language/accounts.mdx]: This API got removed in Stable Cadence, so should be removed from the docs.

**Review by turbolent — COMMENTED** [review]
Great work resolving all those conflicts! 💪

File: runtime/sema/authaccount.cdc [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    ///
+    /// Adds a public key to the account.
+    ///
+    /// The public key must be encoded together with their signature algorithm, hashing algorithm and weight.
+    pub fun addPublicKey(_ publicKey: [UInt8])
+
+    /// **DEPRECATED**: Use `keys.revoke` instead.
+    ///
+    /// Revokes the key at the given index.
+    pub fun removePublicKey(_ index: Int)
```
Comment (turbolent) [inline runtime/sema/authaccount.cdc:47]: These functions got removed in Stable Cadence.

File: encoding/ccf/encode.go (hunk 2 above)
Comment (SupunS) [reply encoding/ccf/encode.go]: Updated: 891935956430045008ec9cb4d3073931f23a1fc8

File: runtime/sema/authaccount.cdc (hunk 5 above)
Comment (SupunS) [reply runtime/sema/authaccount.cdc]: Removed here and from docs: 3d57ec19028c88e752ffb8fff1ce0a4dce49b48f

**Comment by dsainati1** [comment]
Would like to get this merged soon since I'm going to be reworking the entitlement stack.

**Review by turbolent — APPROVED** [review]
👌
